### PR TITLE
Nudging works for private image repositories

### PR DESCRIPTION
**Pull request by rcerven**
### Checklist:
 - PR has reference to the issue(s) it resolves
 - Write / update unit tests
 - Write / update integration (envtest) tests
 - Ensure there is an issue for e2e tests if needed
 - Ensure `make test` passes
 - Ensure test coverage hasn't decreased
 - Test code changes manually
 - Update readme and documentation
 - Write PR description that highlights overall changes and add usage examples if applicable

### CODE REVIEW

File: controllers/component_dependency_update_controller.go [hunk 1]
```diff
@@ -488,6 +539,145 @@ func (r *ComponentDependencyUpdateReconciler) removePipelineFinalizer(ctx contex
 	return ctrl.Result{}, nil
 }
 
+func (r *ComponentDependencyUpdateReconciler) getImageRepositoryCredentials(ctx context.Context, namespace, updatedOutputImage string) (string, string, error) {
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:542]: imo, it's more of search than just get.
Could you provide a good description of the method?

File: controllers/component_dependency_update_controller.go [hunk 2]
```diff
@@ -488,6 +539,145 @@ func (r *ComponentDependencyUpdateReconciler) removePipelineFinalizer(ctx contex
...
+	imageRepositoryUsername, imageRepositoryPassword, err := r.GetCredentialsForImageRepository(ctx, updatedOutputImage, filteredSecretsData)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("unable to find credential for repository %s", updatedOutputImage))
+		return "", "", err
+	}
+	return imageRepositoryUsername, imageRepositoryPassword, nil
+}
+
+// GetCredentialsForImageRepository This method returns credentials for image repository
+func (r ComponentDependencyUpdateReconciler) GetCredentialsForImageRepository(ctx context.Context, outputImage string, imageRepoSecrets []RepositoryCredentials) (string, string, error) {
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:626]: Method name is really similar to the previous one. That's really confusing and until you read the code, it's unclear which one does what. Please rename at least one of them.

File: controllers/component_dependency_update_controller.go [hunk 3]
```diff
@@ -488,6 +539,145 @@ func (r *ComponentDependencyUpdateReconciler) removePipelineFinalizer(ctx contex
...
+
+	imageRepositoryUsername, imageRepositoryPassword, err := r.GetCredentialsForImageRepository(ctx, updatedOutputImage, filteredSecretsData)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("unable to find credential for repository %s", updatedOutputImage))
+		return "", "", err
+	}
+	return imageRepositoryUsername, imageRepositoryPassword, nil
+}
+
+// GetCredentialsForImageRepository This method returns credentials for image repository
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:625]: Description is not great.

File: controllers/component_dependency_update_controller.go (hunk 2 above)
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:626]: Seems that it doesn't need to be a method

File: controllers/component_dependency_update_controller.go [hunk 4]
```diff
@@ -488,6 +539,145 @@ func (r *ComponentDependencyUpdateReconciler) removePipelineFinalizer(ctx contex
...
+			break
+		}
+		partialRepo := strings.Join(repoParts, "/")
+
+		for _, credential := range imageRepoSecrets {
+			credentialRepo := strings.TrimSuffix(credential.RepoName, "/")
+			if partialRepo == credentialRepo {
+				log.Info("partial match found of repository in auth", "repo", partialRepo, "secretName", credential.SecretName)
+				username = credential.UserName
+				password = credential.Password
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:671]: Do we need to check them for emptiness?

File: controllers/component_dependency_update_controller.go [hunk 5]
```diff
@@ -351,16 +387,31 @@ func (r *ComponentDependencyUpdateReconciler) handleCompletedBuild(ctx context.C
 			componentsToUpdate = append(componentsToUpdate, comp)
 		}
 	}
-	var nudgeErr error
 
+	updatedOutputImage := updatedComponent.Spec.ContainerImage
+	imageRepositoryHost := strings.Split(updatedOutputImage, "/")[0]
+
+	imageRepositoryUsername, imageRepositoryPassword, err := r.getImageRepositoryCredentials(ctx, pipelineRun.Namespace, updatedOutputImage)
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:394]: What if image repository is public?

Comment (rcerven) [reply controllers/component_dependency_update_controller.go:394]: even public repository will have push secret

File: controllers/renovate_util.go [hunk 6]
```diff
@@ -287,12 +297,20 @@ func generateRenovateConfigForNudge(target updateTarget, buildResult *BuildResul
 			followTag: "{{.BuiltImageTag}}"
 		  }
 		],
+		hostRules: [
+		  {
+			"matchHost": "{{.ImageRepositoryHost}}",
+			"username": "{{.ImageRepositoryUsername}}",
+			"password": "{{.ImageRepositoryPassword}}"
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:304]: it is already in the secret in the namespace

File: controllers/component_dependency_update_controller.go (hunk 4 above)
Comment (rcerven) [reply controllers/component_dependency_update_controller.go:671]: at this point we didn't find full match

and checking for partial

if there is any auth which has base64 auth without user or password builds would be failing anyway

File: controllers/renovate_util.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/renovate_util.go:304]: I'll create secret and reference it in config

File: controllers/component_dependency_update_controller.go (hunk 2 above)
Comment (rcerven) [reply controllers/component_dependency_update_controller.go:626]: sure

File: controllers/component_dependency_update_controller.go (hunk 4 above)
Comment (rcerven) [reply controllers/component_dependency_update_controller.go:671]: adding

File: controllers/component_dependency_update_controller.go (hunk 2 above)
Comment (rcerven) [reply controllers/component_dependency_update_controller.go]: renamed

File: controllers/component_dependency_update_controller.go (hunk 3 above)
Comment (rcerven) [reply controllers/component_dependency_update_controller.go]: added description

File: controllers/component_dependency_update_controller_test.go [hunk 7]
```diff
@@ -81,6 +81,18 @@ var _ = Describe("Component nudge controller", func() {
 		}
 		createSCMSecret(basicSecretKey, basicSecretData, v1.SecretTypeBasicAuth, map[string]string{})
 
+		imageRepoSecretKey := types.NamespacedName{Name: "dockerconfigjsonsecret", Namespace: UserNamespace}
+		dockerConfigJson := fmt.Sprintf(`{"auths":{"%s":{"auth":"cmVwb191c2VybmFtZTpyZXBvX3Bhc3N3b3JkCg=="}}}`, strings.Split(ComponentContainerImage, ":")[0])
+		imageRepoSecretData := map[string]string{
+			".dockerconfigjson": dockerConfigJson,
+		}
+		createSCMSecret(imageRepoSecretKey, imageRepoSecretData, v1.SecretTypeDockerConfigJson, map[string]string{})
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller_test.go:89]: It looks like a separate test utils function

File: controllers/component_dependency_update_controller_test.go [hunk 8]
```diff
@@ -201,7 +213,50 @@ var _ = Describe("Component nudge controller", func() {
...
 				// check that renovate pipeline run was created
 				renovatePipelinesCreated := getRenovatePipelineRunCount()
+				return renovatePipelinesCreated == 1 && renovateConfigsCreated == 1 && failureCount == 0
+			}, timeout, interval).WithTimeout(ensureTimeout).Should(BeTrue())
+		})
+
+		It("Test build performs nudge on success, image repository partial auth", func() {
+			imageRepoSecretKey := types.NamespacedName{Name: "dockerconfigjsonsecret", Namespace: UserNamespace}
+			// create secret with only partial repository match
+			dockerConfigJson := fmt.Sprintf(`{"auths":{"%s":{"auth":"cmVwb191c2VybmFtZTpyZXBvX3Bhc3N3b3JkCg=="}}}`, strings.Split(ComponentContainerImage, "/")[0])
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller_test.go:223]: Better to construct it from a string, so it's visible what's inside.

**Review by mmorhun — APPROVED** [review]

**Pull request merged**